sha1-checked/src/compress.rs

```diff
@@ -0,0 +1,791 @@
+    ihvout[3] = ihvin[3].wrapping_add(d);
+    ihvout[4] = ihvin[4].wrapping_add(e);
+}
+
+fn recompress_fast_65(
+    ihvin: &mut [u32; 5],
+    ihvout: &mut [u32; 5],
+    me2: &[u32; 80],
+    state: &[u32; 5],
+) {
+    let [mut a, mut b, mut c, mut d, mut e] = state;
+
+    round4_step_bw(b, &mut c, d, e, &mut a, me2[64]);
+    round4_step_bw4(&mut c, &mut d, &mut e, &mut a, &mut b, me2, 63);
+
+    round3_step_bw4(&mut b, &mut c, &mut d, &mut e, &mut a, me2, 59);
+    round3_step_bw4(&mut a, &mut b, &mut c, &mut d, &mut e, me2, 55);
+    round3_step_bw4(&mut e, &mut a, &mut b, &mut c, &mut d, me2, 51);
+    round3_step_bw4(&mut d, &mut e, &mut a, &mut b, &mut c, me2, 47);
+    round3_step_bw4(&mut c, &mut d, &mut e, &mut a, &mut b, me2, 43);
+
+    round2_step_bw4(&mut b, &mut c, &mut d, &mut e, &mut a, me2, 39);
+    round2_step_bw4(&mut a, &mut b, &mut c, &mut d, &mut e, me2, 35);
+    round2_step_bw4(&mut e, &mut a, &mut b, &mut c, &mut d, me2, 31);
+    round2_step_bw4(&mut d, &mut e, &mut a, &mut b, &mut c, me2, 27);
+    round2_step_bw4(&mut c, &mut d, &mut e, &mut a, &mut b, me2, 23);
+
+    round1_step_bw4(&mut b, &mut c, &mut d, &mut e, &mut a, me2, 19);
+    round1_step_bw4(&mut a, &mut b, &mut c, &mut d, &mut e, me2, 15);
+    round1_step_bw4(&mut e, &mut a, &mut b, &mut c, &mut d, me2, 11);
+    round1_step_bw4(&mut d, &mut e, &mut a, &mut b, &mut c, me2, 7);
+    round1_step_bw4(&mut c, &mut d, &mut e, &mut a, &mut b, me2, 3);
+
+    *ihvin = [a, b, c, d, e];
+    [a, b, c, d, e] = *state;
+
+    round4_step(a, &mut b, c, d, &mut e, me2[65]);
+    round4_step(e, &mut a, b, c, &mut d, me2[66]);
+    round4_step(d, &mut e, a, b, &mut c, me2[67]);
+
+    round4_step4(&mut c, &mut d, &mut e, &mut a, &mut b, me2, 68);
+    round4_step4(&mut d, &mut e, &mut a, &mut b, &mut c, me2, 72);
+    round4_step4(&mut e, &mut a, &mut b, &mut c, &mut d, me2, 76);
+
+    ihvout[0] = ihvin[0].wrapping_add(a);
+    ihvout[1] = ihvin[1].wrapping_add(b);
+    ihvout[2] = ihvin[2].wrapping_add(c);
+    ihvout[3] = ihvin[3].wrapping_add(d);
+    ihvout[4] = ihvin[4].wrapping_add(e);
+}
+
+fn recompression_step(
+    step: Testt,
+    ihvin: &mut [u32; 5],
+    ihvout: &mut [u32; 5],
+    me2: &[u32; 80],
+    state: &[u32; 5],
+) {
+    match step {
+        Testt::T58 => {
+            recompress_fast_58(ihvin, ihvout, me2, state);
+        }
+        Testt::T65 => {
+            recompress_fast_65(ihvin, ihvout, me2, state);
+        }
+    }
+}
+
+#[inline(always)]
+fn xor(a: &[u32; 5], b: &[u32; 5]) -> u32 {
+    a[0] ^ b[0] | a[1] ^ b[1] | a[2] ^ b[2] | a[3] ^ b[3] | a[4] ^ b[4]
+}
+
+#[inline]
+pub(super) fn compress(
+    state: &mut [u32; 5],
+    ctx: &mut DetectionState,
+    blocks: &[[u8; BLOCK_SIZE]],
+) {
+    let mut block_u32 = [0u32; BLOCK_SIZE / 4];
+    // since LLVM can't properly use aliasing yet it will make
+    // unnecessary state stores without this copy
+    let mut state_cpy = *state;
```

Comment:
Since we have a lot of code here, it's very unlikely that we will be able to keep the state in registers, so we can work with the state directly without making a copy.

Reply:
removed